cachelib/allocator/CacheAllocator-inl.h

```diff
@@ -1219,23 +1219,52 @@ CacheAllocator<CacheTrait>::findEviction(PoolId pid, ClassId cid) {
   // Keep searching for a candidate until we were able to evict it
   // or until the search limit has been exhausted
   unsigned int searchTries = 0;
-  auto itr = mmContainer.getEvictionIterator();
   while ((config_.evictionSearchTries == 0 ||
-          config_.evictionSearchTries > searchTries) &&
-         itr) {
+          config_.evictionSearchTries > searchTries)) {
     ++searchTries;
 
-    Item* candidate = itr.get();
+    Item* toRecycle = nullptr;
+    Item* candidate = nullptr;
+
+    mmContainer.withEvictionIterator(
+        [this, &candidate, &toRecycle, &searchTries](auto&& itr) {
+          while ((config_.evictionSearchTries == 0 ||
+                  config_.evictionSearchTries > searchTries) &&
+                 itr) {
+            ++searchTries;
+
+            auto* toRecycle_ = itr.get();
+            auto* candidate_ =
+                toRecycle_->isChainedItem()
+                    ? &toRecycle_->asChainedItem().getParentItem(compressor_)
+                    : toRecycle_;
+
+            // make sure no other thead is evicting the item
+            if (candidate_->getRefCount() == 0 && candidate_->markMoving()) {
+              toRecycle = toRecycle_;
+              candidate = candidate_;
+              return;
+            }
+
+            ++itr;
+          }
+        });
+
+    if (!toRecycle)
+      continue;
+
+    XDCHECK(toRecycle);
+    XDCHECK(candidate);
+
     // for chained items, the ownership of the parent can change. We try to
     // evict what we think as parent and see if the eviction of parent
     // recycles the child we intend to.
     auto toReleaseHandle =
-        itr->isChainedItem()
-            ? advanceIteratorAndTryEvictChainedItem(itr)
-            : advanceIteratorAndTryEvictRegularItem(mmContainer, itr);
+        evictNormalItem(*candidate, true /* skipIfTokenInvalid */);
+    auto ref = candidate->unmarkMoving();
 
-    if (toReleaseHandle) {
-      if (toReleaseHandle->hasChainedItem()) {
+    if (toReleaseHandle || ref == 0u) {
```

Comment:
Why can we rely on just `ref == 0` to proceed with eviction?

Without a `toReleaseHandle`, it could mean we didn't remove the item (or its parent item) from access-container. It's not safe to evict it since it can still be accessed.

Reply:
Please see my answer below.

I actually think that we might destroy the `toReleaseHandle` first (unconditionally) and then only rely on `ref==0u`. `toReleaseHandle` will not actually release anything on destruction since we have the moving bit set. And if `unmarkMoving()` will return `0` we know that, either we or some other thread, managed to remove the item from AC and MMContainer and no one is using the item. I'll check this.